File: 18-2.py
#!/usr/bin/env python

import numpy
from collections import deque


class Map:
    def __init__(self, max_x, max_y, max_z):
        self.size_x = max_x + 3
        self.size_y = max_y + 3
        self.size_z = max_z + 3
        self.map = numpy.zeros((self.size_x, self.size_y, self.size_z), dtype=numpy.uint8)

    def flood_fill(self):
        queue = deque([[0, 0, 0]])
        while len(queue) > 0:
            cell = queue.popleft()
            if self.map[cell[0], cell[1], cell[2]] != 0:
                continue

            self.map[cell[0], cell[1], cell[2]] = 2

            if cell[0] > 0 and self.map[cell[0] - 1, cell[1], cell[2]] == 0:
                queue.append([cell[0] - 1, cell[1], cell[2]])
            if cell[0] < self.size_x - 1 and self.map[cell[0] + 1, cell[1], cell[2]] == 0:
                queue.append([cell[0] + 1, cell[1], cell[2]])
            if cell[1] > 0 and self.map[cell[0], cell[1] - 1, cell[2]] == 0:
                queue.append([cell[0], cell[1] - 1, cell[2]])
            if cell[1] < self.size_y - 1 and self.map[cell[0], cell[1] + 1, cell[2]] == 0:
                queue.append([cell[0], cell[1] + 1, cell[2]])
            if cell[2] > 0 and self.map[cell[0], cell[1], cell[2] - 1] == 0:
                queue.append([cell[0], cell[1], cell[2] - 1])
            if cell[2] < self.size_z - 1 and self.map[cell[0], cell[1], cell[2] + 1] == 0:
                queue.append([cell[0], cell[1], cell[2] + 1])


    def count_exposed_faces(self):
        count = 0
        for x in range(self.size_x):
            for y in range(self.size_y):
                for z in range(self.size_z):
                    if self.map[x, y, z] == 1:
                        if x == 0 or self.map[x - 1, y, z] == 2:
                            count += 1
                        if x == self.size_x - 1 or self.map[x + 1, y, z] == 2:
                            count += 1
                        if y == 0 or self.map[x, y - 1, z] == 2:
                            count += 1
                        if y == self.size_y - 1 or self.map[x, y + 1, z] == 2:
                            count += 1
                        if z == 0 or self.map[x, y, z - 1] == 2:
                            count += 1
                        if z == self.size_z - 1 or self.map[x, y, z + 1] == 2:
                            count += 1
        return count


lines = [list(map(lambda c: int(c), line.strip().split(","))) for line in open('18.input').readlines()]

max_x = max([x for x, _, _ in lines])
max_y = max([y for _, y, _ in lines])
max_z = max([z for _, _, z in lines])

obj = Map(max_x, max_y, max_z)

for coords in lines:
    obj.map[coords[0], coords[1], coords[2]] = 1

obj.flood_fill()
print(obj.count_exposed_faces())

File: 18-2.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  def initialize(max_x, max_y, max_z)
    @size_x = max_x + 3
    @size_y = max_y + 3
    @size_z = max_z + 3
    @map = Numo::UInt8.zeros(@size_x, @size_y, @size_z)
  end

  def [](x, y, z)
    @map[x + 1, y + 1, z + 1]
  end

  def []=(x, y, z, val)
    @map[x + 1, y + 1, z + 1] = val
  end

  def flood_fill!
    queue = [[0, 0, 0]]
    until queue.empty?
      cell = queue.shift
      next if @map[cell[0], cell[1], cell[2]] != 0

      @map[cell[0], cell[1], cell[2]] = 2
      queue << [cell[0] - 1, cell[1], cell[2]] if cell[0] > 0 and @map[cell[0] - 1, cell[1], cell[2]].zero?
      queue << [cell[0] + 1, cell[1], cell[2]] if cell[0] < @size_x - 1 and @map[cell[0] + 1, cell[1], cell[2]].zero?
      queue << [cell[0], cell[1] - 1, cell[2]] if cell[1] > 0 and @map[cell[0], cell[1] - 1, cell[2]].zero?
      queue << [cell[0], cell[1] + 1, cell[2]] if cell[1] < @size_y - 1 and @map[cell[0], cell[1] + 1, cell[2]].zero?
      queue << [cell[0], cell[1], cell[2] - 1] if cell[2] > 0 and @map[cell[0], cell[1], cell[2] - 1].zero?
      queue << [cell[0], cell[1], cell[2] + 1] if cell[2] < @size_z - 1 and @map[cell[0], cell[1], cell[2] + 1].zero?
    end
  end

  def count_exposed_faces
    count = 0
    (@size_x - 1).times do |x|
      (@size_y - 1).times do |y|
        (@size_z - 1).times do |z|
          next unless @map[x + 1, y + 1, z + 1] == 1

          count += 1 if @map[x + 1 - 1, y + 1, z + 1] == 2
          count += 1 if @map[x + 1 + 1, y + 1, z + 1] == 2
          count += 1 if @map[x + 1, y + 1 - 1, z + 1] == 2
          count += 1 if @map[x + 1, y + 1 + 1, z + 1] == 2
          count += 1 if @map[x + 1, y + 1, z + 1 - 1] == 2
          count += 1 if @map[x + 1, y + 1, z + 1 + 1] == 2
        end
      end
    end
    count
  end
end

input = File.read('18.input').lines.map(&:strip).map { |l| l.split(',').map(&:to_i) }

max_x = input.map(&:first).max
max_y = input.map { |c| c[1] }.max
max_z = input.map(&:last).max

map = Map.new(max_x, max_y, max_z)

input.each do |coords|
  map[coords[0], coords[1], coords[2]] = 1
end

map.flood_fill!

print map.count_exposed_faces, "\n"
